### DISCUSSION TITLE
405 Not Allowed Error

### DISCUSSION THREAD

**Discussion by strawbs1989** [community/community#124546] [Programming Help]
### Select Topic Area

Question

### Body

why when testing a project the site gives me a 405 Not Allowed 

i'v only got this code 

index.html: `<!DOCTYPE HTML>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Song Request Form</title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/water.css@2/out/water.css">
</head>
<body>
    <img src="https://coolvibes-reloaded.com/img/favicon.png" width="100px" height="100px">
    <center>
        <form id="reqForm" action="saverequest.php" method="POST">
            <label for="from">From:</label><br>
            <input type="text" name="from" id="from" required><br>
            
            <label for="name">To:</label><br>
            <input type="text" name="name" id="name" pattern="[a-z]{4,9}" required><br>
            <p>To must be lowercase and 4-9 characters in length.</p>
            
            <label for="djs">Choose A DJ:</label><br>
            <select name="djs" id="djs" required>
                <option value="dj kat">Classic Kat</option>
                <option value="dj strawbs">Strawbs</option>
                <option value="dj dale">Dale</option>
                <option value="dj jon">Jon</option>
                <option value="dj simon">Simon</option>
                <option value="dj twist">Christina</option>
            </select><br><br>
            
            <label for="message">Message:</label><br>
            <input type="text" name="message" id="message" required /><br>
            
            <label for="songrequest">Song Request:</label><br>
            <input type="text" name="songrequest" placeholder="Ex: Plan B - She Said" id="songrequest" required /><br><br>
            
            <button type="submit">Submit</button>
        </form>
    </center>
</body>
</html>
`
SaveRequest.php: `<?php
if ($_SERVER["REQUEST_METHOD"] == "POST") {
    $from = htmlspecialchars($_POST['from']);
    $to = htmlspecialchars($_POST['name']);
    $djs = htmlspecialchars($_POST['djs']);
    $message = htmlspecialchars($_POST['message']);
    $songrequest = htmlspecialchars($_POST['songrequest']);

    $data = "From: $from, To: $to, DJ: $djs, Message: $message, Song Request: $songrequest\n";
    file_put_contents('jaystest/requests.txt', $data, FILE_APPEND);

    header('Location: jaystest/success.html');
    exit();
} else {
    echo "Invalid request method.";
}
?>
`
success.html: `<!DOCTYPE HTML>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Success</title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/water.css@2/out/water.css">
</head>
<body>
    <h1>Request Successful</h1>
    <p>Your song request and dedication has been submitted successfully!</p>
    <a href="jaystest/index.html">Submit another request</a>
</body>
</html>
`
viewrequest.php: `<!DOCTYPE HTML>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>View Requests</title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/water.css@2/out/water.css">
</head>
<body>
    <h1>Song Requests and Dedications</h1>
    <ul>
        <?php
        if (file_exists('requests.txt')) {
            $requests = file('jaystest/requests.txt', FILE_IGNORE_NEW_LINES);
            foreach ($requests as $request) {
                echo "<li>$request</li>";
            }
        } else {
            echo "<li>No requests found.</li>";
        }
        ?>
    </ul>
</body>
</html>
`

why is it giving me a 405 Not Allowed

**Comment by mranv** [comment]
The 405 Not Allowed error typically occurs when the server refuses to accept the HTTP method used in the request. This can happen if the server is configured to only allow certain HTTP methods (e.g., GET, POST) for a particular URL or resource.

Based on the code you provided, it seems that you are not handling the HTTP method correctly in your PHP script (`SaveRequest.php`). The code you shared is:

```php
$data = "From: $from, To: $to, DJ: $djs, Message: $message, Song Request: $songrequest\n";
file_put_contents('jaystest/requests.txt', $data, FILE_APPEND);
header('Location: jaystest/success.html');
exit();
```

This code assumes that the request method is always POST, which is typically used for form submissions. However, if the request method is GET (e.g., when you access the page directly in the browser), the script will not handle it correctly, leading to the 405 Not Allowed error.

To fix this issue, you need to check the request method and handle it appropriately. Here's an example of how you can modify your `SaveRequest.php` script:

```php
<?php
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    // Handle form submission
    $from = $_POST['from'];
    $to = $_POST['to'];
    $djs = $_POST['djs'];
    $message = $_POST['message'];
    $songrequest = $_POST['songrequest'];

    $data = "From: $from, To: $to, DJ: $djs, Message: $message, Song Request: $songrequest\n";
    file_put_contents('jaystest/requests.txt', $data, FILE_APPEND);
    header('Location: jaystest/success.html');
    exit();
} else {
    // Handle invalid request method
    http_response_code(405); // Set the appropriate HTTP status code
    echo "Invalid request method.";
}
?>
```

In this modified code, we first check the `REQUEST_METHOD` using `$_SERVER['REQUEST_METHOD']`. If the request method is POST, it means the form was submitted, and we handle the form data as before. However, if the request method is not POST (e.g., GET), we send a 405 Not Allowed status code using `http_response_code(405)` and display an appropriate error message.

By adding this check and handling the request method correctly, your script should no longer return the 405 Not Allowed error when accessed directly in the browser or through an invalid request method.

**Comment by liker912** [comment]
Hi @strawbs1989 What is the action url in the form? And what method (POST, GET, etc) ?
